### PR TITLE
Add fast path for BGEN 1.2 dosages

### PR DESCRIPTION
**Pull request by cseed**
In the UKBBv2 case.

Depends on https://github.com/hail-is/hail/pull/1957

### CODE REVIEW

File: src/main/scala/is/hail/io/bgen/BgenRecord.scala [hunk 1]
```diff
@@ -116,13 +117,46 @@ class Bgen12GenotypeIterator(a: Array[Byte],
 
   private val pa = new BGen12ProbabilityArray(a, nSamples, nGenotypes, nBitsPerProb)
 
+  def sampleMissing(s: Int): Boolean = (a(8 + s) & 0x80) != 0
```
Comment (jbloom22) [inline src/main/scala/is/hail/io/bgen/BgenRecord.scala]: consider changing to `isMissingSample` to indicate Boolean value, pairs will with `missingSamples`

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 2]
```diff
@@ -210,6 +212,43 @@ class LoadBgenSuite extends SparkSuite {
     packedInput
   }
 
+  @Test def testDosage() {
+    val sc = hc.sc
+    hc.indexBgen("src/test/resources/example.8bits.bgen")
+    val vds = hc.importBgen("src/test/resources/example.8bits.bgen"
+      // , nPartitions = Some(8)
```
Comment (jbloom22) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: Consider removing comment and testing with multiple partitions, move `,` to previous line

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 3]
```diff
@@ -210,6 +212,43 @@ class LoadBgenSuite extends SparkSuite {
...
+      , nPartitions = Some(1)
+    )
+
+    val mask = Array.fill(vds.nSamples)(true)
+    val incomplete = Array(13, 172, 273, 319, 441)
```
Comment (jbloom22) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: use `incompleteSampleIndices` to be consistent with `completeSampleIndices` below

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 4]
```diff
@@ -210,6 +212,43 @@ class LoadBgenSuite extends SparkSuite {
...
+    val maskBc = sc.broadcast(mask)
+
+    val completeSampleIndices = (0 until vds.nSamples)
+      .filter(mask)
+      .toArray
+    assert(completeSampleIndices.length == nKept)
+    val completeSampleIndicesBc = sc.broadcast(completeSampleIndices)
+
+    vds.rdd.foreachPartition { it =>
+      val missingSamples = new ArrayBuilder[Int]
```
Comment (jbloom22) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: Add `()`

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 5]
```diff
@@ -210,6 +212,43 @@ class LoadBgenSuite extends SparkSuite {
...
+      val dosages2 = new DenseVector[Double](nKept)
+
+      it.foreach { case (v, (va, gs)) =>
+        val dosages1 = RegressionUtils.dosages(gs, nKept, maskBc.value)
```
Comment (jbloom22) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: `dosages1` => `dosages`
`dosages2` => `dosagesBgen12`

File: src/test/scala/is/hail/io/LoadBgenSuite.scala [hunk 6]
```diff
@@ -210,6 +212,43 @@ class LoadBgenSuite extends SparkSuite {
...
+      // , nPartitions = Some(8)
+      , nPartitions = Some(1)
+    )
+
+    val mask = Array.fill(vds.nSamples)(true)
+    val incomplete = Array(13, 172, 273, 319, 441)
+    for (i <- incomplete)
+      mask(i) = false
+
+    val nKept = vds.nSamples - incomplete.length
```
Comment (jbloom22) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: suggestion: `nKept` => `nComplete`

Comment (cseed) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: This is the name you use in RegressionUtils.dosages.  Left it.

File: src/test/scala/is/hail/io/LoadBgenSuite.scala (hunk 5 above)
Comment (cseed) [inline src/test/scala/is/hail/io/LoadBgenSuite.scala]: I liked mine better and left it.

**Comment by cseed** [comment]
Addressed comments.  Back to you.

**Pull request merged**